### PR TITLE
Fix make_simplified_union interaction with Any

### PR DESCRIPTION
**Pull request by ddfisher**
make_simplified_union had two incorrect interactions with Any that this
fixes:
1) Any unions containing Any were simplified to Any, which is incorrect.
2) Any classes that inherited from Any would be removed from the union
by the subclass simplification (because subclasses of Any are
considered subclasses of all other classes due to subtype/compatible
with confusion).

Note that `Union`s call make_union and not make_simplified_union, so
this doesn't change their behavior directly (unless they interact with
something else which causes them to be simplified, like being part of
an `Optional`).

### CODE REVIEW

File: test-data/unit/check-optional.test [hunk 1]
```diff
@@ -428,3 +428,10 @@ x + 1
 [out]
 main:2: note: In module imported here:
 tmp/a.py:3: error: Unsupported left operand type for + (some union)
+
+[case testOptionalFallbackToNone]
+from typing import Any, Optional
+class C(Any):
+    pass
+x = None  # type: Optional[C]
+x.foo()  # E: Some element of union has no attribute "foo"
```
Comment (JukkaL) [inline test-data/unit/check-optional.test]: Ideas for additional test cases:

* Test case for using `isinstance` with something like `Optional[C]` (`C` has an `Any` base class), and test how the conditional type binder works in that case.
* Similar to above (`isinstance`), but with `Union[Any, int]` where there is no `Any` base class.
* Test accessing an attribute of `Union[Any, int]` where there is no `Any` base class.

File: mypy/types.py [hunk 2]
```diff
@@ -858,17 +858,16 @@ def make_simplified_union(items: List[Type], line: int = -1, column: int = -1) -
...
         removed = set()  # type: Set[int]
         for i, ti in enumerate(items):
             if i in removed: continue
             # Keep track of the truishness info for deleted subtypes which can be relevant
             cbt = cbf = False
             for j, tj in enumerate(items):
-                if i != j and is_subtype(tj, ti):
+                tj_is_anylike = (isinstance(tj, AnyType) or
+                                 (isinstance(tj, Instance) and tj.type.fallback_to_any))
+                if i != j and is_subtype(tj, ti) and not tj_is_anylike:
```
Comment (ddfisher) [inline mypy/types.py]: Actually, I've changed my mind.  I don't feel strongly about this, but I think the conditional is moderately more readable if `tj_is_anylike` comes after the subtype check.  `tj_is_anylike` isn't true often enough for the perf difference to matter.

If you still prefer it otherwise, I'll switch it.

File: test-data/unit/check-optional.test (hunk 1 above)
Comment (ddfisher) [inline test-data/unit/check-optional.test]: Maybe I'm missing something, but those test cases seem completely unrelated to this issue?

File: mypy/types.py (hunk 2 above)
Comment (gvanrossum) [inline mypy/types.py]: I don't feel strongly either. *I* don't have any other objections, but I'll wait with merging this until it's clear what Jukka wanted. Or Jukka can merge.

**Comment by ddfisher** [comment]
@JukkaL is there anything more you'd like to see here before merging?

File: test-data/unit/check-optional.test (hunk 1 above)
Comment (JukkaL) [inline test-data/unit/check-optional.test]: This introduces a new sort of union type (one with both `Any` and non-`Any` types), and some other code might make assumptions about this and not work correctly with them. I haven't looked at the relevant code, but it's probably as easy to write the tests than to manually validate that everything is okay. In particular, I remember than union-related operations were implemented in a slightly sketchy fashion and thus might have problems (though they are probably fine)... Alternatively, it would be equally fine to have unit tests that just verify that operations on unions still do the right thing -- `restrict_subtype_away` comes to mind, in particular.

**Comment by JukkaL** [comment]
About my comment: 

Using a real subtype check would certainly be a better fix than what I suggested. However, the current PR seems incomplete in that it special cases `Any` but doesn't address concerns with other types such as `List[Any]` vs. `List[int]`. My suggestion is to modify `make_simplified_union` to also handle the latter case so that it always returns `List[Any]` instead of arbitrarily returning `List[Any]` or `List[int]`, via taking a join of the union items that are compatible with each other. If you are going to implement the actual subtype check in a follow-up PR soon, this is not important, and the suggestion is optional in any case. If you don't plan to implement the real subtype check soon, I'd suggest adding a comment about why the current fix is still incomplete.

The real subtype check might be somewhat non-trivial to implement in a way that handles all the corner cases correctly, so I think that it would be reasonable to make a temporary fix that is still not quite correct but less incorrect than what we currently have, as the required incremental change would be small :-)

About subtypes vs compatibility:

I don't see how `is_subtype` is broken, other than being wrongly named, based on the current terminology. If we rename it to `is_compatible`, is it still broken? There currently is no implementation of actual subtype checking (other than `is_proper_subtype`, which isn't complete). Modifying `is_subtype` to perform a subtype check is not the right thing to do -- we need a new, separate operation that will again be called `is_subtype` (perhaps it can share some code with `is_compatible`). Almost all current uses of `is_subtype` are correct, as far as I know, and the calls would just need to be updated to call `is_compatible` instead (or whatever it will be called). 

Historically, I used 'is subtype of' to mean what is currently understood as 'is compatible with', and 'is proper subtype of' for what is currently understood as 'is subtype of'. In retrospect that was a bad decision, and I can well see how it's confusing to use 'subtype' to mean something that is quite different from the textbook definition.

File: test-data/unit/check-optional.test (hunk 1 above)
Comment (ddfisher) [inline test-data/unit/check-optional.test]: This actually _doesn't_ introduce a new Union type.  Explicitly writing `Union[A, B]` called `make_union`, not `make_simplified_union`, so these types already exist.

File: mypy/types.py [hunk 3]
```diff
@@ -858,17 +858,18 @@ def make_simplified_union(items: List[Type], line: int = -1, column: int = -1) -
...
         removed = set()  # type: Set[int]
         for i, ti in enumerate(items):
             if i in removed: continue
             # Keep track of the truishness info for deleted subtypes which can be relevant
             cbt = cbf = False
             for j, tj in enumerate(items):
-                if i != j and is_subtype(tj, ti):
+                # Attempt to only combine true subtypes by avoiding types containing Any.
+                # TODO: Properly exclude generics and functions.
+                tj_is_anylike = (isinstance(tj, AnyType) or
```
Comment (JukkaL) [inline mypy/types.py]: What if there are multiple `Any` types in the union to simplify? I think that we should remove duplicates away at least, perhaps by using `is_any_type`. So `Union[Any, Any, int]` would be simplified into `Union[Any, int]`.

File: test-data/unit/check-optional.test (hunk 1 above)
Comment (JukkaL) [inline test-data/unit/check-optional.test]: There was already an issue open for one of the bugs I was worried about: #1720

**Comment by gvanrossum** [comment]
@JukkaL @ddfisher are you shooting to get this into the 0.4.5 release?

**Comment by JukkaL** [comment]
As we are pretty close to release, I'm fine with postponing this until the next release. Then we could perhaps release a larger set of fixes to unions. But if @ddfisher wants to get this in it doesn't look very risky once the duplicate `Any` issue is resolved. However, I'd like to have this tested on real-world code, as the union type implementation has a bunch of issues (unrelated to this PR) and this might cause some of them to become more visible.

**Comment by gvanrossum** [comment]
OK let's postpone it. (David agreed off-line.)

**Comment by JukkaL** [comment]
@ddfisher Can you look at simplifying away duplicate `Any` values in unions? It's worth adding a test case for this.

**Comment by gvanrossum** [comment]
Ping

**Comment by ddfisher** [comment]
To do this right, I'm going to fix up `is_proper_subtype`.

**Comment by rwbarton** [comment]
BTW, can someone explain what "proper subtype" is supposed to mean in the context of mypy? I never understood it.

**Comment by ddfisher** [comment]
Here's my understanding: in mypy, "subtype" means "compatible with" and "proper subtype" means "subtype".

**Comment by rwbarton** [comment]
Okay, that makes sense. I saw the terminology "compatible with" elsewhere in mypy and thought it was a separate, third concept but I guess it isn't.

**Comment by JukkaL** [comment]
I need something similar for another thing I'm working on, so I'm probably going to do another PR that replaces this. I'll keep this open until I have the other PR ready.

**Comment by JukkaL** [comment]
#2714 is an updated version of this PR -- it handles some additional cases and adds tests.
